File: Algorithm Problems/Maximum_Subarray.cpp
//Maximum Subarray
//Date 19 Aug 2020
//Source https://leetcode.com/problems/maximum-subarray/

class Solution {
public:
    int maxSubArray(vector<int>& nums) {
		int dp[nums.size()];
		dp[0]=nums[0];
		int res=nums[0];
        for(int i=1;i<nums.size();i++){
			int s= (dp[i-1]>0? dp[i-1]:0);
			dp[i]= nums[i]+ s;
			res= max(res, dp[i]);
		}
		return res;
    }
};

File: Algorithm Problems/Maximum_Subarray.py
# 1 July 2021

class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        if len(nums)==1:
            return nums[0]
        dp=[]
        dp.append(nums[0])
        for i in range(1,len(nums)):
            if dp[i-1]<0:
                dp.append(nums[i])
            else:
                dp.append(nums[i]+dp[i-1])
        return max(dp)
